From b19baf99a51a3aebdc0360eaeddce9d433cabc63 Mon Sep 17 00:00:00 2001
From: Juliusz Sosinowicz <juliusz@wolfssl.com>
Date: Fri, 24 Jan 2025 10:13:35 +0100
Subject: [PATCH] Support for wolfSSL

Compile wolfSSL with
```
./configure --enable-all
make
make install
```

Compile libimobiledevice with
```
patch -p1 < <path/to/patch/file>
./autogen.sh --enable-wolfssl
make
make install
```
---
 3rd_party/libsrp6a-sha512/Makefile.am |  3 +++
 common/Makefile.am                    |  3 +++
 common/userpref.c                     | 23 ++++++++++++++++++-
 configure.ac                          | 33 +++++++++++++++++++++++++++
 cython/Makefile.am                    |  2 ++
 tools/idevicebackup.c                 |  3 +++
 6 files changed, 66 insertions(+), 1 deletion(-)

diff --git a/3rd_party/libsrp6a-sha512/Makefile.am b/3rd_party/libsrp6a-sha512/Makefile.am
index c349d8c..22987df 100644
--- a/3rd_party/libsrp6a-sha512/Makefile.am
+++ b/3rd_party/libsrp6a-sha512/Makefile.am
@@ -6,6 +6,8 @@ AM_CPPFLAGS = \
         -Wno-incompatible-pointer-types
 
 AM_CFLAGS = -DHAVE_CONFIG_H
+if HAVE_WOLFSSL
+AM_CFLAGS += -DOPENSSL=1 $(wolfssl_CFLAGS)
 if HAVE_OPENSSL
 AM_CFLAGS += -DOPENSSL=1 $(openssl_CFLAGS)
 else
@@ -17,6 +19,7 @@ AM_CFLAGS += -DMBEDTLS=1 $(mbedtls_CFLAGS)
 endif
 endif
 endif
+endif
 
 noinst_LTLIBRARIES = libsrp6a-sha512.la
 
diff --git a/common/Makefile.am b/common/Makefile.am
index 1a90571..f698bf6 100644
--- a/common/Makefile.am
+++ b/common/Makefile.am
@@ -6,10 +6,12 @@ AM_CFLAGS = \
 	$(GLOBAL_CFLAGS) \
 	$(libusbmuxd_CFLAGS) \
 	$(libplist_CFLAGS) \
+	$(limd_glue_CFLAGS) \
 	$(libgnutls_CFLAGS) \
 	$(libtasn1_CFLAGS) \
 	$(libgcrypt_CFLAGS) \
 	$(openssl_CFLAGS) \
+	$(wolfssl_CFLAGS) \
 	$(LFS_CFLAGS)
 
 AM_LDFLAGS = \
@@ -19,6 +21,7 @@ AM_LDFLAGS = \
 	$(libtasn1_LIBS) \
 	$(libgcrypt_LIBS) \
 	$(openssl_LIBS) \
+	$(wolfssl_LIBS) \
 	${libpthread_LIBS}
 
 noinst_LTLIBRARIES = libinternalcommon.la
diff --git a/common/userpref.c b/common/userpref.c
index 11e28ba..29e4a05 100644
--- a/common/userpref.c
+++ b/common/userpref.c
@@ -43,7 +43,7 @@
 #include <openssl/rsa.h>
 #include <openssl/x509.h>
 #include <openssl/x509v3.h>
-#if OPENSSL_VERSION_NUMBER < 0x1010000fL || \
+#if (!defined(HAVE_WOLFSSL) && OPENSSL_VERSION_NUMBER < 0x1010000fL) || \
 	(defined(LIBRESSL_VERSION_NUMBER) && (LIBRESSL_VERSION_NUMBER < 0x20700000L))
 #define X509_set1_notBefore X509_set_notBefore
 #define X509_set1_notAfter X509_set_notAfter
@@ -368,6 +368,10 @@ static int X509_add_ext_helper(X509 *cert, int nid, char *value)
 	X509_EXTENSION *ex;
 	X509V3_CTX ctx;
 
+#ifdef HAVE_WOLFSSL
+	memset(&ctx, 0, sizeof(ctx));
+#endif
+
 	/* No configuration database */
 	X509V3_set_ctx_nodb(&ctx);
 
@@ -457,7 +461,14 @@ userpref_error_t pair_record_generate_keys_and_certs(plist_t pair_record, key_da
 	{
 		/* set serial number */
 		ASN1_INTEGER* sn = ASN1_INTEGER_new();
+#ifdef HAVE_WOLFSSL
+		/* wolfSSL doesn't permit using 0 for serial numbers, in accordance with
+		 * RFC 5280:
+		 * https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.2. */
+		ASN1_INTEGER_set(sn, 1);
+#else
 		ASN1_INTEGER_set(sn, 0);
+#endif
 		X509_set_serialNumber(root_cert, sn);
 		ASN1_INTEGER_free(sn);
 
@@ -487,7 +498,12 @@ userpref_error_t pair_record_generate_keys_and_certs(plist_t pair_record, key_da
 	{
 		/* set serial number */
 		ASN1_INTEGER* sn = ASN1_INTEGER_new();
+#ifdef HAVE_WOLFSSL
+		/* See note above on 0 serial numbers. */
+		ASN1_INTEGER_set(sn, 2);
+#else
 		ASN1_INTEGER_set(sn, 0);
+#endif
 		X509_set_serialNumber(host_cert, sn);
 		ASN1_INTEGER_free(sn);
 
@@ -574,7 +590,12 @@ userpref_error_t pair_record_generate_keys_and_certs(plist_t pair_record, key_da
 	if (pubkey && dev_cert) {
 		/* generate device certificate */
 		ASN1_INTEGER* sn = ASN1_INTEGER_new();
+#ifdef HAVE_WOLFSSL
+		/* See note above on 0 serial numbers. */
+		ASN1_INTEGER_set(sn, 3);
+#else
 		ASN1_INTEGER_set(sn, 0);
+#endif
 		X509_set_serialNumber(dev_cert, sn);
 		ASN1_INTEGER_free(sn);
 		X509_set_version(dev_cert, 2);
diff --git a/configure.ac b/configure.ac
index 8ba094a..c608fd6 100644
--- a/configure.ac
+++ b/configure.ac
@@ -160,6 +160,14 @@ AC_SUBST([CYTHON_SUB])
 
 default_openssl=yes
 
+AC_ARG_ENABLE([wolfssl],
+            [AS_HELP_STRING([--enable-wolfssl],
+            [Use wolfSSL crypto])],
+            [use_wolfssl=$enableval],
+            [use_wolfssl=no])
+if test "x$use_wolfssl" == "xyes"; then
+  default_openssl=no
+fi
 AC_ARG_WITH([mbedtls],
             [AS_HELP_STRING([--without-mbedtls],
             [Do not look for mbedtls])],
@@ -208,6 +216,30 @@ if test "x$use_mbedtls" == "xyes"; then
   ssl_provider="MbedTLS";
   ssl_requires=""
   AC_SUBST(ssl_requires)
+elif test "x$use_wolfssl" = "xyes"; then
+  AC_DEFINE(HAVE_WOLFSSL, 1, [Define if you have wolfSSL support])
+  AC_DEFINE(HAVE_OPENSSL, 1, [Defined to use wolfSSL compatibility layer])
+  PKG_CHECK_MODULES(
+    [wolfssl],
+    [wolfssl],
+    [],
+    [AC_MSG_ERROR([Could not find wolfSSL.])]
+  )
+  PKG_CHECK_VAR(
+    [wolfssl_includedir],
+    [wolfssl],
+    [includedir],
+    [],
+    [AC_MSG_ERROR([Could not find wolfSSL includedir variable.])]
+  )
+  wolfssl_CFLAGS="${wolfssl_CFLAGS} -DWOLFSSL_USE_OPTIONS_H -DEXTERNAL_OPTS_OPENVPN -I${wolfssl_includedir}/wolfssl"
+  ssl_lib_CFLAGS="$wolfssl_CFLAGS"
+  ssl_lib_LIBS="$wolfssl_LIBS"
+  AC_SUBST(ssl_lib_CFLAGS)
+  AC_SUBST(ssl_lib_LIBS)
+  ssl_provider="wolfSSL"
+  ssl_requires="wolfssl"
+  AC_SUBST(ssl_requires)
 else
   if test "x$use_openssl" == "xyes"; then
     pkg_req_openssl="openssl >= 0.9.8"
@@ -248,6 +280,7 @@ else
 fi
 AM_CONDITIONAL(HAVE_MBEDTLS, test "x$use_mbedtls" == "xyes")
 AM_CONDITIONAL(HAVE_OPENSSL, test "x$use_openssl" == "xyes")
+AM_CONDITIONAL(HAVE_WOLFSSL, test "x$use_wolfssl" == "xyes")
 AM_CONDITIONAL(HAVE_GCRYPT, test "x$use_gnutls" == "xyes")
 
 AC_ARG_ENABLE([wireless-pairing],
diff --git a/cython/Makefile.am b/cython/Makefile.am
index 3577c4e..afc7461 100644
--- a/cython/Makefile.am
+++ b/cython/Makefile.am
@@ -6,6 +6,7 @@ AM_CFLAGS = \
 	$(libgnutls_CFLAGS) \
 	$(libtasn1_CFLAGS) \
 	$(openssl_CFLAGS) \
+	$(wolfssl_CFLAGS) \
 	$(libplist_CFLAGS) \
 	$(LFS_CFLAGS) \
 	$(PTHREAD_CFLAGS)
@@ -14,6 +15,7 @@ AM_LIBS = \
 	$(libgnutls_LIBS) \
 	$(libtasn1_LIBS) \
 	$(openssl_LIBS) \
+	$(wolfssl_CFLAGS) \
 	$(libplist_LIBS) \
 	$(PTHREAD_LIBS)
 
diff --git a/tools/idevicebackup.c b/tools/idevicebackup.c
index 1684666..30ab2cd 100644
--- a/tools/idevicebackup.c
+++ b/tools/idevicebackup.c
@@ -32,6 +32,9 @@
 #include <stdlib.h>
 #include <signal.h>
 #include <getopt.h>
+#ifdef HAVE_WOLFSSL
+#include <openssl/ssl.h>
+#endif
 #if defined(HAVE_OPENSSL)
 #include <openssl/sha.h>
 #elif defined(HAVE_GNUTLS)
-- 
2.34.1

